crypto/rsa/rsa_gen.c

```diff
@@ -253,15 +254,26 @@ static int rsa_builtin_keygen(RSA *rsa, int bits, int primes, BIGNUM *e_value,
              * 1. 1024-bit factors are effcient when using 3072 and 4096-bit key
              * 2. stay the same logic with normal 2-prime key
              */
+            bitse -= bitsr[i];
+            if (!BN_GENCB_call(cb, 2, n++))
+                goto err;
             if (primes > 4) {
                 if (bitst < 0x9)
                     adj++;
                 else
                     adj--;
+            } else if (i == primes - 1 && retries == 4) {
+                /*
+                 * re-generate all primes from scratch, mainly used
+                 * in 4 prime case to avoid long loop. Max retry times
+                 * is set to 4.
+                 */
+                i = -1;
+                retries = 0;
```

Comment:
I would suggest to move that to where the adj = 0 is now,
so that failures on previous primes do not accumulate.

Reply:
... sorry for so many comments ;-)